worker/peergrouper/desired.go

```diff
@@ -397,88 +399,128 @@ func (p *peerGroupChanges) getMachinesVoting() {
 
-	if addr != p.members[id].Address {
-		p.members[id].Address = addr
-		p.isChanged = true
-	}
-	return nil
-}
-
-// updateAddressFromInternal attempts to update the member with a cloud-local
-// address from the machine.
+// updateAddressesFromInternal attempts to update each member with a
+// cloud-local address from the machine.
 // If there is a single cloud local address available, it is used.
 // If there are multiple addresses, then a check is made to ensure that:
 //   - the member was previously in the replica-set and;
 //   - the previous address used for replication is still available.
 // If the check is satisfied, then a warning is logged and no change is made.
 // Otherwise an error is returned to indicate that a HA space must be
-// configured in order to proceed.
-func (p *peerGroupChanges) updateAddressFromInternal(id string, info *peerGroupInfo) error {
-	member := p.members[id]
-	hostPorts := info.machines[id].GetPotentialMongoHostPorts(info.mongoPort)
-	addrs := network.SelectInternalHostPorts(hostPorts, false)
-
-	// This should not happen because SelectInternalHostPorts will choose a
-	// public address when there are no cloud-local addresses.
-	// Zero addresses would mean the machine would not be accessible at all.
-	// We ignore this outcome and leave the address alone.
-	if len(addrs) == 0 {
-		return nil
-	}
+// configured in order to proceed. Such machines have their status set to
+// indicate that they require intervention.
+func (p *peerGroupChanges) updateAddressesFromInternal(info *peerGroupInfo) error {
+	var multipleAddresses []string
 
-	if len(addrs) == 1 {
-		addr := addrs[0]
-		logger.Debugf("machine %q selected address %q by scope from %v", id, addr, hostPorts)
+	for id := range p.members {
+		m := info.machines[id]
+		hostPorts := m.GetPotentialMongoHostPorts(info.mongoPort)
+		addrs := network.SelectInternalHostPorts(hostPorts, false)
+
+		// This should not happen because SelectInternalHostPorts will choose a
+		// public address when there are no cloud-local addresses.
+		// Zero addresses would mean the machine is completely inaccessible.
+		// We ignore this outcome and leave the address alone.
+		if len(addrs) == 0 {
+			continue
+		}
 
-		if member.Address != addr {
-			member.Address = addr
-			p.isChanged = true
+		// Unique address; we can use this for Mongo peer communication.
+		member := p.members[id]
+		if len(addrs) == 1 {
+			addr := addrs[0]
+			logger.Debugf("machine %q selected address %q by scope from %v", id, addr, hostPorts)
+
+			if member.Address != addr {
+				member.Address = addr
+				p.isChanged = true
+			}
+			continue
+		}
+
+		// Multiple potential Mongo addresses.
+		// Checks are required in order to use it as a peer.
+		unchanged := false
+		if _, ok := info.recognised[id]; ok {
+			for _, addr := range addrs {
+				if member.Address == addr {
+					logger.Warningf("%s\npreserving member with unchanged address %q", multiAddressMessage, addr)
```

Comment:
this feels worthy of setting status. Do you agree? I could be convinced otherwise, but it does feel like a case of "we're still guessing, and things don't seem right".

Reply:
Should the status itself remain as _started_, with such information in the message field?

Reply:
At this stage it remains without a status message.

The reason for this is that this condition could be present in a successful peer group change, and the current implementation would reset the status message anyway.

Working around this would require some specific indicator (maybe in MachineTracker?) that indicates, this machine is part of successful peer-group determination, but do not blank the status message for this specific situation.